frontends/p4/sideEffects.cpp

```diff
@@ -61,13 +72,36 @@ struct EvaluationOrder {
     }
 };
 
+/** @brief Replace expressions containing constructor or method invocations
+ * with temporaries.
+ * 
+ * The EvaluationOrder field accumulates lists of the temporaries introduced in
+ * this way, as well as assignment statements that assign the original
+ * expressions to the appropriate temporaries.
+ *
+ * The DoSimplifyExpressions pass will later insert the declarations and
+ * assignments above the expression.
+ *
+ * @pre An up-to-date ReferenceMap and TypeMap.
+ *
+ * @post The RefenceMap and TypeMap are updated to reflect changes to the IR.
+ * This includes extra information stored in the TypeMap, specifically
+ * left-value-ness and compile-time-contant-ness.
+ */
 class DismantleExpression : public Transform {
+    // TODO: It may be worth refactoring DismantleExpression to purely use
```

Comment:
We actually want to precisely control the visit order, to reflect the order that the evaluation should be performed. I don't want to rely on the visitor order, which is not well-defined. This is the only way to really control the precise order: manual calls to visit and prune. See the compiler-design document too.